crates/dataflow-jit/src/codegen/intrinsics.rs

```diff
@@ -296,6 +330,94 @@ unsafe extern "C" fn dataflow_jit_string_with_capacity(capacity: usize) -> ThinS
     ThinStr::with_capacity(capacity)
 }
 
+macro_rules! timestamp_intrinsics {
+    ($($name:ident => $expr:expr),+ $(,)?) => {
+        paste::paste! {
+            $(
+                unsafe extern "C" fn [<dataflow_jit_timestamp_ $name>](millis: i64) -> i64 {
+                    if let LocalResult::Single(timestamp) = Utc.timestamp_millis_opt(millis) {
+                        let expr: fn(DateTime<Utc>) -> i64 = $expr;
+                        expr(timestamp)
+                    } else {
+                        tracing::error!(
+                            "failed to create timestamp from {millis} in {}",
+                            stringify!([<dataflow_jit_timestamp_ $name>]),
+                        );
+                        0
+                    }
+                }
+            )+
+        }
+    }
+}
+
+// TODO: Some of these really return u32 or i32
+timestamp_intrinsics! {
+    millennium => |time| ((time.year() + 999) / 1000) as i64,
+    century => |time| ((time.year() + 99) / 100) as i64,
+    decade => |time| (time.year() / 10) as i64,
+    year => |time| time.year() as i64,
+    iso_year => |time| time.iso_week().year() as i64,
+    quarter => |time| (time.month0() / 3 + 1) as i64,
+    month => |time| time.month() as i64,
+    week => |time| time.iso_week().week() as i64,
+    day => |time| time.month() as i64,
+    hour => |time| time.hour() as i64,
+    minute => |time| time.minute() as i64,
+    second => |time| time.second() as i64,
+    millisecond => |time| (time.second() * 1000 + time.timestamp_subsec_millis()) as i64,
+    microsecond => |time| (time.second() * 1_000_000 + time.timestamp_subsec_micros()) as i64,
+    day_of_week => |time| (time.weekday().num_days_from_sunday() as i64) + 1,
+    iso_day_of_week => |time| (time.weekday().num_days_from_monday() as i64) + 1,
+    day_of_year => |time| time.ordinal() as i64,
+    floor_week => |time| {
```

Comment:
This does not look correct. 
The definition of the week number is insane:
"ISO 8601 includes the ISO week-date system, a numbering system for weeks. Each week is associated with the year in which Thursday occurs. For example, week 1 of 2004 (2004W01) covered Monday, 29 December 2003 to Sunday, 4 January 2004. European countries/regions typically use this style of numbering. Non-European countries/regions typically don't use it." 
This comes from https://learn.microsoft.com/en-us/sql/t-sql/functions/datepart-transact-sql?view=sql-server-ver16
I really would prefer to use the existing Chrono libraries instead of implementing this by hand.
All of these functions already exist in sqllib/src/timestamp.rs, it would be great to reuse that code.

Reply:
sorry, this comment is not valid; floow_week is right. it applies to week.